**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 2**
Pandas para el análisis de datos en Python

---

*   NOMBRE: ELVAN LOPEZ RODRIGUEZ
*   MATRÍCULA: A01841568


---

En esta actividad trabajarás con el archivo `cleaned_weather.csv`, un extracto del conjunto de datos meteorológicos a lo largo de todo el año 2020 en una estación del Instituto *Max Planck* (Alemania) disponible en Kaggle.

Los datos meteorológicos fueron registrados cada 10 minutos e incluyen los siguientes indicadores:

*   `timestamp`: Fecha y hora de la observación.
*   `p`: Presión atmosférica en milibares (mbar)
*   `T`: Temperatura del aire en grados Celsius (°C)
*   `Tpot`: Temperatura potencial en Kelvin (K)
*   `rh`: Humedad relativa en porcentaje (%)
*   `VPact`: Presión real de vapor en milibares (mbar)
*   `sh`: Humedad específica en gramos por kilogramo (g/kg)
*   `H2OC`: Concentración de vapor de agua en milimoles por mol (mmol/mol) de aire seco
*   `rho`: Densidad del aire en gramos por metro cúbico (g/m³)
*   `wv`: Velocidad del viento en metros por segundo (m/s)
*   `wd`: Dirección del viento en grados (°)
*   `rain`: Precipitación total en milímetros (mm)
*   `raining`: Duración de la lluvia en segundos (s)

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


In [375]:
import pandas as pd

1.	Descarga el archivo: `cleaned_weather.csv` y guarda, en un dataframe (`weather_df`), todos sus registros.
- Determina la dimensionalidad del dataframe y obtén los identificadores de columnas.
- Renombra las columnas para facilitar la interpretación de los indicadores en los ejercicios siguientes:
  - `medicion, presion_atmosferica, temperatura_celsius, temperatura_kelvin, humedad_relativa, presion_vapor, humedad_especifica, concentracion_vapor, densidad_aire, velocidad_viento, direccion_viento, precipitacion_total, duracion_lluvia`
- Muestra los primeros y los últimos 5 registros.
- ¿Hay valores faltantes en el dataframe?





In [376]:
df = pd.read_csv("cleaned_weather.csv")
print("Tamaño del dataframe:\n")
print(f"{df.shape[0]} filas, {df.shape[1]} columnas")


Tamaño del dataframe:

52696 filas, 13 columnas


In [377]:
print("\n------------------------------------------------------------------------")
col = list(df.columns)
print("Columnas en dataframe:\n")
for i in range(len(df.columns)):
    print(i + 1, ") ", col[i])


------------------------------------------------------------------------
Columnas en dataframe:

1 )  timestamp
2 )  p
3 )  T
4 )  Tpot
5 )  rh
6 )  VPact
7 )  sh
8 )  H2OC
9 )  rho
10 )  wv
11 )  wd
12 )  rain
13 )  raining


In [378]:
print("\n------------------------------------------------------------------------")
df.columns = [
    "medicion",
    "presion_atmosferica",
    "temperatura_celsius",
    "temperatura_kelvin",
    "humedad_relativa",
    "presion_vapor",
    "humedad_especifica",
    "concentracion_vapor",
    "densidad_aire",
    "velocidad_viento",
    "direccion_viento",
    "precipitacion_total",
    "duracion_lluvia",
]

col = list(df.columns)
print("Columnas en dataframe renombradas:\n")
for i in range(len(df.columns)):
    print(i + 1, ") ", col[i])



------------------------------------------------------------------------
Columnas en dataframe renombradas:

1 )  medicion
2 )  presion_atmosferica
3 )  temperatura_celsius
4 )  temperatura_kelvin
5 )  humedad_relativa
6 )  presion_vapor
7 )  humedad_especifica
8 )  concentracion_vapor
9 )  densidad_aire
10 )  velocidad_viento
11 )  direccion_viento
12 )  precipitacion_total
13 )  duracion_lluvia


In [379]:
print("\n------------------------------------------------------------------------")

print("Cinco primeros registros:\n")
df.head(5)


------------------------------------------------------------------------
Cinco primeros registros:



,medicion,presion_atmosferica,temperatura_celsius,temperatura_kelvin,humedad_relativa,presion_vapor,humedad_especifica,concentracion_vapor,densidad_aire,velocidad_viento,direccion_viento,precipitacion_total,duracion_lluvia
0,01/01/20 0:10,1008.89,0.71,273.18,86.1,5.54,3.42,5.49,1280.62,1.02,224.3,0.0,0
1,01/01/20 0:20,1008.76,0.75,273.22,85.2,5.49,3.39,5.45,1280.33,0.43,206.8,0.0,0
2,01/01/20 0:30,1008.66,0.73,273.21,85.1,5.48,3.39,5.43,1280.29,0.61,197.1,0.0,0
3,01/01/20 0:40,1008.64,0.37,272.86,86.3,5.41,3.35,5.37,1281.97,1.11,206.4,0.0,0
4,01/01/20 0:50,1008.61,0.33,272.82,87.4,5.47,3.38,5.42,1282.08,0.49,209.6,0.0,0


In [380]:
print("\n------------------------------------------------------------------------")

print("Cinco últimos registros:\n")
df.tail(5)


------------------------------------------------------------------------
Cinco últimos registros:



,medicion,presion_atmosferica,temperatura_celsius,temperatura_kelvin,humedad_relativa,presion_vapor,humedad_especifica,concentracion_vapor,densidad_aire,velocidad_viento,direccion_viento,precipitacion_total,duracion_lluvia
52691,31/12/20 23:20,978.32,2.28,277.16,80.0,5.76,3.67,5.89,1234.61,0.73,180.6,0.0,0
52692,31/12/20 23:30,978.30,2.13,277.01,83.1,5.92,3.77,6.05,1235.20,0.43,174.0,0.0,0
52693,31/12/20 23:40,978.26,1.99,276.88,82.2,5.80,3.69,5.93,1235.82,0.38,248.9,0.0,0
52694,31/12/20 23:50,978.26,2.07,276.95,81.4,5.77,3.68,5.90,1235.49,0.57,196.6,0.0,0
52695,01/01/21 0:00,978.24,2.01,276.89,82.4,5.82,3.71,5.95,1235.71,0.57,221.3,0.0,0


In [381]:
print("\n------------------------------------------------------------------------")

print("Sumatoria de valores nulos por columna:\n")
print(df.isna().sum())
print(f"\nExisten {df.isna().sum().sum()} valores nulos")


------------------------------------------------------------------------
Sumatoria de valores nulos por columna:

medicion               0
presion_atmosferica    0
temperatura_celsius    0
temperatura_kelvin     0
humedad_relativa       0
presion_vapor          0
humedad_especifica     0
concentracion_vapor    0
densidad_aire          0
velocidad_viento       0
direccion_viento       0
precipitacion_total    0
duracion_lluvia        0
dtype: int64

Existen 0 valores nulos


2. Determina el tipo de datos que tienen las columnas.
- Cambia la columna `medicion` a datetime utilizando el formato: `%d/%m/%y %H:%M`
- Obtén dos columnas adicionales separando `medicion` en `fecha` y `hora`

In [382]:
print("\n------------------------------------------------------------------------")

print(f"Tipo de dato actual de la columna 'medicion': '{df['medicion'].dtypes}'")

df["medicion"] = pd.to_datetime(df["medicion"], format="%d/%m/%y %H:%M")

print(
    f"Después del cambio de formato, la columna 'medicion' tiene un tipo:\
    '{df['medicion'].dtypes}'"
)



------------------------------------------------------------------------
Tipo de dato actual de la columna 'medicion': 'str'
Después del cambio de formato, la columna 'medicion' tiene un tipo:    'datetime64[us]'


In [383]:
print("\n------------------------------------------------------------------------")

print(f"Creación de la columna 'fecha'")

df["fecha"] = df["medicion"].dt.date
df["fecha"].head()


------------------------------------------------------------------------
Creación de la columna 'fecha'


0    2020-01-01
1    2020-01-01
2    2020-01-01
3    2020-01-01
4    2020-01-01
Name: fecha, dtype: object

In [384]:
print("\n------------------------------------------------------------------------")

print(f"Creación de la columna 'hora'")

df["hora"] = df["medicion"].dt.time
df["hora"].head()


------------------------------------------------------------------------
Creación de la columna 'hora'


0    00:10:00
1    00:20:00
2    00:30:00
3    00:40:00
4    00:50:00
Name: hora, dtype: object

3. Determina la cantidad de valores únicos de cada columna.
- La columna `medicion` tiene un valor menos que el número total de filas del dataframe, lo que indica que hay una fecha duplicada. Identifica cuál es.
- Comprueba si las demás columnas también contienen los mismos valores; si es así, elimina  uno de los registros duplicados.


In [385]:
print("\n------------------------------------------------------------------------")

print(
    "Revisar ambos registros duplicados (sólo en 'medicion') y "
    "comparar sus otras columnas para saber si hay duplicados:"
)
regDupEnMedicion = df[df.duplicated(subset=["medicion"], keep=False)]

valorRegDupEnMedicion = regDupEnMedicion.duplicated().sum()
print(f"\tExisten {valorRegDupEnMedicion} registro(s) duplicado(s)\n")

if valorRegDupEnMedicion > 0:
    print("Tamaño actual del dataframe:")
    print(f"{df.shape[0]} filas, {df.shape[1]} columnas\n")

    print(
        f"\t\tComo hay {valorRegDupEnMedicion} registro(s) duplicado(s), "
        "lo(s) eliminaremo(s) dejando sólamente el primero"
    )

    indiceRegABorrar = df[df["medicion"].duplicated()]

    print(
        f"\t\tSe elimnará(n) el(los) siguiente(s) registro(s): {list(indiceRegABorrar.index)}"
    )

    df = df.drop_duplicates(subset=["medicion"], keep="first")

print("\nNuevo tamaño del dataframe:")
print(f"{df.shape[0]} filas, {df.shape[1]} columnas")


------------------------------------------------------------------------
Revisar ambos registros duplicados (sólo en 'medicion') y comparar sus otras columnas para saber si hay duplicados:
	Existen 1 registro(s) duplicado(s)

Tamaño actual del dataframe:
52696 filas, 15 columnas

		Como hay 1 registro(s) duplicado(s), lo(s) eliminaremo(s) dejando sólamente el primero
		Se elimnará(n) el(los) siguiente(s) registro(s): [19044]

Nuevo tamaño del dataframe:
52695 filas, 15 columnas


4. Imprime nuevamente las cantidad de valores únicos por columna.
- ¿Por qué son 144 horas diferentes?
- La cantidad de combinaciones únicas de fecha y hora no coincide con el número total de mediciones. Aunque no haya valores NaN, esto no garantiza que no se hayan omitido registros. Identifica las fechas en las que faltan mediciones.
- Una vez concluido el análisis de mediciones faltantes, elimina todos los registros que no correspondan al año 2020 y la columna `medicion`.


In [386]:
print("\n------------------------------------------------------------------------")

print(f"Registros totales: {len(df)}\n")

print("Registros únicos por columna:\n")
print(df.nunique())


------------------------------------------------------------------------
Registros totales: 52695

Registros únicos por columna:

medicion               52695
presion_atmosferica     5052
temperatura_celsius     3680
temperatura_kelvin      3836
humedad_relativa        5494
presion_vapor           2059
humedad_especifica      1337
concentracion_vapor     2087
densidad_aire          14826
velocidad_viento         977
direccion_viento        8099
precipitacion_total       42
duracion_lluvia           61
fecha                    367
hora                     144
dtype: int64


In [387]:
print("\n------------------------------------------------------------------------")

print(
    "Parece ser que las mediciones fueron hechas cada 10min, ello implicaría"
    "24 * 6 = 144 medicionenes (y por lo tanto horas diferentes) al día."
)
print(df["hora"].head(50))


------------------------------------------------------------------------
Parece ser que las mediciones fueron hechas cada 10min, ello implicaría24 * 6 = 144 medicionenes (y por lo tanto horas diferentes) al día.
0     00:10:00
1     00:20:00
2     00:30:00
3     00:40:00
4     00:50:00
5     01:00:00
6     01:10:00
7     01:20:00
8     01:30:00
9     01:40:00
10    01:50:00
11    02:00:00
12    02:10:00
13    02:20:00
14    02:30:00
15    02:40:00
16    02:50:00
17    03:00:00
18    03:10:00
19    03:20:00
20    03:30:00
21    03:40:00
22    03:50:00
23    04:00:00
24    04:10:00
25    04:20:00
26    04:30:00
27    04:40:00
28    04:50:00
29    05:00:00
30    05:10:00
31    05:20:00
32    05:30:00
33    05:40:00
34    05:50:00
35    06:00:00
36    06:10:00
37    06:20:00
38    06:30:00
39    06:40:00
40    06:50:00
41    07:00:00
42    07:10:00
43    07:20:00
44    07:30:00
45    07:40:00
46    07:50:00
47    08:00:00
48    08:10:00
49    08:20:00
Name: hora, dtype: object


In [388]:
print("\n------------------------------------------------------------------------")

cantidadMedicionesXFecha = (
    df.groupby("fecha")["hora"].count().reset_index(name="mediciones")
)

# 2. Filter using the new column name
fechasMedicionesIncompletas = cantidadMedicionesXFecha[
    cantidadMedicionesXFecha["mediciones"] < 144
]
print("\nFechas con mediciones incompletas:")
print(fechasMedicionesIncompletas)



------------------------------------------------------------------------

Fechas con mediciones incompletas:
          fecha  mediciones
0    2020-01-01         143
149  2020-05-29         135
366  2021-01-01           1


5. Obtén un dataframe `indicadores_diarios` que sólo almacene el mayor valor de las mediciones por fecha para responder:
- ¿Cuál fue la máxima precipitación y a qué fecha pertenece?
- ¿Cuáles fueron los tres días con las velocidades de viento más altas registradas?
- ¿Cuántos días rebasaron el 90% de humedad relativa y los 30° celsius?

6. La función `describe()` resume múltiples estadísticas descriptivas. Utiliza esta función para explorar y responder las preguntas sobre las variables en el dataframe de `indicadores_diarios `:
- ¿Cuál es la desviación estándar de la concentración de vapor y qué indica?
- ¿Cuál es la mediana de la presión atmosférica y qué significa?
- ¿Cuál es el tercer cuartil de la duración de la lluvia y cómo se interpreta?

7.	Dibuja un histograma para cada una de las columnas de `indicadores_diarios`.

8. Utiliza el dataframe de indicadores diarios para crear un nuevo dataframe `indicadores_mensuales`. Para ello:
- Crea una nueva columna `mes` a partir de la fecha que se encuentra en el índice.
- Haciendo uso de `groupby()`, calcula los promedios mensuales y conserva únicamente las columnas `humedad_relativa`, `temperatura_celsius` y `velocidad_viento` en el nuevo dataframe.

9. Construye el mismo dataframe mensual, pero esta vez utilizando la función `pivot_table()`.
- Añade la columna `indice_calor` usando la fórmula de Thom (*Temperature - Humidity Index*, TDI):
> `temperatura_celsius + 0.33 x humedad_relativa - 0.7 x velocidad_viento - 4`

10. Recuerda que de las cuatro funciones estudiadas para manipular la estructura del dataframe:
- `melt/pivot` hacen las transformaciones de manera controlada, es decir puedes
definir por medio de sus parámetros qué variables quedarán como índices, columnas y valores.
- `stack/unstack` la conversión se aplica siempre sobre los niveles inferiores de index/columns.
- Convierte el dataframe del ejercicio anterior a formato largo usando: `melt()` y `stack()` y comenta las diferencias.

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

---